bpf/process/bpf_generic_lsm.c

```diff
@@ -0,0 +1,162 @@
+	__uint(max_entries, 1);
+	__type(key, int);
+	__type(value, struct filter_map_value);
+} filter_map SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct event_config);
+} config_map SEC(".maps");
+
+#ifdef __LARGE_BPF_PROG
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct msg_data);
+} data_heap SEC(".maps");
+#define data_heap_ptr &data_heap
+#else
+#define data_heap_ptr 0
+#endif
+
+static struct generic_maps maps = {
+	.heap = (struct bpf_map_def *)&process_call_heap,
+	.calls = (struct bpf_map_def *)&lsm_calls,
+	.config = (struct bpf_map_def *)&config_map,
+	.filter = (struct bpf_map_def *)&filter_map,
+	.override = (struct bpf_map_def *)&override_tasks,
+};
+
+#define MAIN "lsm/generic_lsm"
+
+__attribute__((section((MAIN)), used)) int
+generic_lsm_event(struct pt_regs *ctx)
+{
+	return generic_start_process_filter(ctx, &maps);
+}
+
+__attribute__((section("lsm/0"), used)) int
+generic_lsm_setup_event(void *ctx)
+{
+	return generic_process_event_and_setup(
+		ctx, (struct bpf_map_def *)&process_call_heap,
+		(struct bpf_map_def *)&lsm_calls,
+		(struct bpf_map_def *)&config_map,
+		(struct bpf_map_def *)data_heap_ptr);
+}
+
+__attribute__((section("lsm/1"), used)) int
+generic_lsm_process_event(void *ctx)
+{
+	return generic_process_event(ctx,
+				     (struct bpf_map_def *)&process_call_heap,
+				     (struct bpf_map_def *)&lsm_calls,
+				     (struct bpf_map_def *)&config_map,
+				     (struct bpf_map_def *)data_heap_ptr);
+}
+
+__attribute__((section("lsm/2"), used)) int
+generic_lsm_process_filter(void *ctx)
+{
+	int ret;
+
+	ret = generic_process_filter((struct bpf_map_def *)&process_call_heap,
+				     (struct bpf_map_def *)&filter_map);
+	if (ret == PFILTER_CONTINUE)
+		tail_call(ctx, &lsm_calls, TAIL_CALL_FILTER);
+	else if (ret == PFILTER_ACCEPT)
+		tail_call(ctx, &lsm_calls, 0);
+	return PFILTER_REJECT;
+}
+
+__attribute__((section("lsm/3"), used)) int
+generic_lsm_filter_arg(void *ctx)
+{
+	return filter_read_arg(ctx, (struct bpf_map_def *)&process_call_heap,
+			       (struct bpf_map_def *)&filter_map,
+			       (struct bpf_map_def *)&lsm_calls,
+			       (struct bpf_map_def *)&config_map,
+			       true);
+}
+
+__attribute__((section("lsm/4"), used)) int
+generic_lsm_actions(void *ctx)
+{
+	generic_actions(ctx, &maps);
+
+	// If NoPost action is set, check for Override action here
+	__u64 id = get_current_pid_tgid();
+	__s32 *error;
+
+	error = map_lookup_elem(&override_tasks, &id);
+	if (!error)
+		return 0;
+
+	map_delete_elem(&override_tasks, &id);
+	return (long)*error;
+}
+
+__attribute__((section("lsm/5"), used)) int
+generic_lsm_output(void *ctx)
+{
+	__u64 id = get_current_pid_tgid();
+	__s32 *error;
+
+	generic_output(ctx, (struct bpf_map_def *)&process_call_heap, MSG_OP_GENERIC_LSM);
+
+	error = map_lookup_elem(&override_tasks, &id);
```

Comment:
Hmm, what am I missing here, in the previous tail call you could delete this entry? so making this check here will fail which will lead to return 0 ?  (I didn't check this part of code lately, so just dropping a comment...)

Reply:
The case you mentioned is only works when we set `NoPost` Action.

https://github.com/cilium/tetragon/blob/2a31728a71b689554fc1fca4573945908cae20e6/bpf/process/types/basic.h#L2329-L2332

Reply:
I think the reason is that the override is done either in generic_lsm_actions or in generic_lsm_output depending on whether there's output or not.. but I think it'd good to have it in separate function like (untested):

```
diff --git a/bpf/process/bpf_generic_lsm.c b/bpf/process/bpf_generic_lsm.c
index 4a8736c180a3..1e80afcfa307 100644
--- a/bpf/process/bpf_generic_lsm.c
+++ b/bpf/process/bpf_generic_lsm.c
@@ -128,11 +128,8 @@ generic_lsm_filter_arg(void *ctx)
                               true);
 }
 
-__attribute__((section("lsm/4"), used)) int
-generic_lsm_actions(void *ctx)
+FUNC_INLINE int try_override(void *ctx)
 {
-       generic_actions(ctx, &maps);
-
        // If NoPost action is set, check for Override action here
        __u64 id = get_current_pid_tgid();
        __s32 *error;
@@ -145,18 +142,18 @@ generic_lsm_actions(void *ctx)
        return (long)*error;
 }
 
+__attribute__((section("lsm/4"), used)) int
+generic_lsm_actions(void *ctx)
+{
+       generic_actions(ctx, &maps);
+
+       // If NoPost action is set, check for Override action here
+       return try_override(ctx);
+}
+
 __attribute__((section("lsm/5"), used)) int
 generic_lsm_output(void *ctx)
 {
-       __u64 id = get_current_pid_tgid();
-       __s32 *error;
-
        generic_output(ctx, (struct bpf_map_def *)&process_call_heap, MSG_OP_GENERIC_LSM);
-
-       error = map_lookup_elem(&override_tasks, &id);
-       if (!error)
-               return 0;
-
-       map_delete_elem(&override_tasks, &id);
-       return (long)*error;
+       return try_override(ctx);
 }

```

Reply:
Yes, it's a good idea!